From 77ccbbf779bb9b51e86587afd3822a3dbef07c78 Mon Sep 17 00:00:00 2001
From: Qihui Zhao <qihuizhao@google.com>
Date: Thu, 25 Jul 2024 15:59:39 +0000
Subject: [PATCH] Fix Chrome crash by returning 0 on kUnsupported IBAN country

This CL fixes crash when it might happen that the user input an IBAN
value which is not in the supported country list, thus cause
GetLengthOfIbanCountry() crash, it fixes the crash by returning 0 length
which is also a valid way to validate the IBAN value.

(cherry picked from commit eb8416da8eb76b1a31b8ceb7dab3ea5eb3d930c0)

Bug: 353606603
Change-Id: I5d64d2845690064f33966526e168dde7f0dd7482
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5714855
Code-Coverage: findit-for-me@appspot.gserviceaccount.com <findit-for-me@appspot.gserviceaccount.com>
Reviewed-by: Stephen McGruer <smcgruer@chromium.org>
Commit-Queue: Qihui Zhao <qihuizhao@google.com>
Reviewed-by: Olivia Saul <jsaul@google.com>
Cr-Original-Commit-Position: refs/heads/main@{#1328997}
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5739413
Reviewed-by: Florian Leimgruber <fleimgruber@google.com>
Cr-Commit-Position: refs/branch-heads/6533@{#1812}
Cr-Branched-From: 7e0b87ec6b8cb5cb2969e1479fc25776e582721d-refs/heads/main@{#1313161}
---

diff --git a/components/autofill/core/browser/data_model/iban.cc b/components/autofill/core/browser/data_model/iban.cc
index 0bbb113..e31c301 100644
--- a/components/autofill/core/browser/data_model/iban.cc
+++ b/components/autofill/core/browser/data_model/iban.cc
@@ -134,9 +134,10 @@
   }
 
   // IBAN length must match the length of IBANs in the country the IBAN is from.
-  size_t iban_value_length = GetLengthOfIbanCountry(
-      GetIbanSupportedCountry(base::UTF16ToUTF8(iban_value.substr(0, 2))));
-  if (iban_value_length == 0 || iban_value_length != iban_value.length()) {
+  const std::string country_code = base::UTF16ToUTF8(iban_value.substr(0, 2));
+  if (!IsIbanApplicableInCountry(country_code) ||
+      GetLengthOfIbanCountry(GetIbanSupportedCountry(country_code)) !=
+          iban_value.length()) {
     return false;
   }
 
